nni/retiarii/graph.py

```diff
@@ -0,0 +1,654 @@
+        self.name: str = name or f'_generated_{graph_id}'
+
+        self.input_node: Node = Node(self, _InputPseudoUid, '_inputs', _IOPseudoOperation('_inputs'), _internal=True)
+        self.output_node: Node = Node(self, _OutputPseudoUid, '_outputs', _IOPseudoOperation('_outputs'), _internal=True)
+        self.hidden_nodes: List[Node] = []
+
+        self.edges: List[Edge] = []
+
+    def __repr__(self):
+        return f'Graph(id={self.id}, name={self.name}, ' + \
+            f'input_names={self.input_node.operation.io_names}, ' + \
+            f'output_names={self.output_node.operation.io_names}, ' + \
+            f'num_hidden_nodes={len(self.hidden_nodes)}, num_edges={len(self.edges)})'
+
+    @property
+    def nodes(self) -> List['Node']:
+        return [self.input_node, self.output_node] + self.hidden_nodes
+
+    def _add_input(self, input_name) -> None:
+        if self.input_node.operation.io_names is None:
+            self.input_node.operation.io_names = [input_name]
+        else:
+            self.input_node.operation.io_names.append(input_name)
+
+    def _add_output(self, output_name) -> None:
+        if self.output_node.operation.io_names is None:
+            self.output_node.operation.io_names = [output_name]
+        else:
+            self.output_node.operation.io_names.append(output_name)
+
+    @overload
+    def add_node(self, name: str, operation: Operation) -> 'Node': ...
+    @overload
+    def add_node(self, name: str, type_name: str, parameters: Dict[str, Any] = {}) -> 'Node': ...
+
+    def add_node(self, name, operation_or_type, parameters={}):
+        if isinstance(operation_or_type, Operation):
+            op = operation_or_type
+        else:
+            op = Operation.new(operation_or_type, parameters, name)
+        return Node(self, self.model._uid(), name, op, _internal=True)._register()
+
+    @overload
+    def insert_node_on_edge(self, edge: 'Edge', name: str, operation: Operation) -> 'Node': ...
+    @overload
+    def insert_node_on_edge(self, edge: 'Edge', name: str, type_name: str, parameters: Dict[str, Any] = {}) -> 'Node': ...
+
+    def insert_node_on_edge(self, edge, name, operation_or_type, parameters={}) -> 'Node':
+        if isinstance(operation_or_type, Operation):
+            op = operation_or_type
+        else:
+            op = Operation.new(operation_or_type, parameters, name)
+        new_node = Node(self, self.model._uid(), name, op, _internal=True)._register()
+        # update edges
+        self.add_edge((edge.head, edge.head_slot), (new_node, None))
+        self.add_edge((new_node, None), (edge.tail, edge.tail_slot))
+        self.del_edge(edge)
+        return new_node
+
+    # mutation
+    def add_edge(self, head: Tuple['Node', Optional[int]], tail: Tuple['Node', Optional[int]]) -> 'Edge':
+        assert head[0].graph is self and tail[0].graph is self
+        return Edge(head, tail, _internal=True)._register()
+
+    def del_edge(self, edge: 'Edge') -> None:
```

Comment:
Please sync the name. Either rename this to `remove_edge`, or rename `Edge.remove` to `del`.